extensions/target-specific/chromium/blink/fragment-tree/fragment-tree.js

```diff
@@ -0,0 +1,78 @@
+Loader.OnLoad(function() {
+    function fragmentToConcreteType(fragment) {
+        return fragment.f("type_").desc().then((type) => {
+            switch (type) {
+                case "kFragmentBox":
+                case "kFragmentRenderedLegend":
+                  return fragment.as(Chromium.RendererProcessType("blink::NGPhysicalBoxFragment"));
+                case "kFragmentLineBox":
+                    return fragment.as(Chromium.RendererProcessType("blink::NGPhysicalLineBoxFragment"));
+                case "kFragmentText":
+                    return fragment.as(Chromium.RendererProcessType("blink::NGPhysicalTextFragment"));
+                default:
+                return fragment;
+            }
+        });
+    }
+
+    NGPhysicalFragmentTree = {
+        Tree: new DbgObjectTree.DbgObjectTreeReader(),
+        Renderer: new DbgObjectTree.DbgObjectRenderer(),
+        InterpretAddress: function(address) {
+            var voidObject = DbgObject.create(Chromium.RendererProcessType("void"), address);
+            if (!voidObject.isNull()) {
+                return fragmentToConcreteType(voidObject.as(Chromium.RendererProcessType("blink::NGPhysicalFragment")));
+            } else {
+                return DbgObject.NULL;
+            }
+        },
+        GetRoots: function() {
+            return BlinkHelpers.GetDocuments()
+            .then((documents) => {
+                if (documents.length == 0) {
+                    var errorMessage = ErrorMessages.CreateErrorsList("No documents found.") +
+                        ErrorMessages.CreateErrorReasonsList(ErrorMessages.WrongDebuggee("the Chromium renderer process"),
+                            "The debuggee has been broken into prior to <i>g_frame_map</i> being populated.",
+                            ErrorMessages.SymbolsUnavailable) +
+                        "You may still specify a blink::NGPhysicalFragment explicitly.";
+                    return Promise.reject(errorMessage);
+                } else {
+                    // We have to get the child of the root layout object (the LayoutView) because LayoutView never has a fragment,
+                    return Promise.map(documents, (document) => document.F("node_layout_data_").f("layout_object_").vcast().f("children_.first_child_").vcast().f("cached_layout_result_.ptr_.physical_fragment_.ptr_"));
+                }
+            }, (error) => {
+                var errorMessage = ErrorMessages.CreateErrorsList(error) +
+                    ErrorMessages.CreateErrorReasonsList(ErrorMessages.WrongDebuggee("the Chromium renderer process"), ErrorMessages.SymbolsUnavailable);
+                return Promise.reject(errorMessage);
+            });
+        },
+        DefaultTypes: [Chromium.RendererProcessType("blink::NGPhysicalFragment")]
+    };
+
+    NGPhysicalFragmentTree.Tree.addChildren(Chromium.RendererProcessType("blink::NGPhysicalFragment"), (fragment) => {
+        return fragment.array("children_");

```

Comment:
What's the reason for including NGLinkStorage as part of the tree view? If it isn't necessary, it might be cleaner to just see a tree of NGPhysicalFragments.

Reply:
NGLinkStorage stores the offset, which I thought was useful to have; and with the addNameRenderer call this doesn't look too bad, I thought (see also the screenshot on Slack #layoutng)

Reply:
Yeah, I saw that. I thought it wasn't immediately obviously looking at the screenshot that the intermediary pieces were NGLinkStorage. Perhaps we should include the NGLinkStorage type name alongside the offset? Another option could be to display the offset alongside the NGPhysicalFragment types that are children of NGLinkStorage (though I don't know if there's a way to know this, since NGPhysicalFragment doesn't point back to its containing NGLinkStorage)?

Reply:
Done. Yes, I don't know how to avoid showing the NGLinkStorage intermixed though I'd love to find a way to do so.

Reply:
Another way is to add a 'synthetic' child to the tree, in the same way we add "Shadow Tree" or "Content Fragment" labels in the DOMTree extension. You could either make the NGLinkStorage node synthetic, or add synthetic children for each offset under NGLinkStorage and append the child physical fragment to that.

It's really up to you though if you want to keep it as is or go the synthetic child way. Just giving you anther option...